builtin/providers/openstack/resource_openstack_images_image_v2.go

```diff
@@ -0,0 +1,488 @@
+				Type:         schema.TypeString,
+				Optional:     true,
+				ForceNew:     false,
+				ValidateFunc: resourceImagesImagevalidateVisibility,
+				Default:      images.ImageVisibilityPrivate,
+			},
+		},
+	}
+}
+
+func resourceImagesImagevalidateVisibility(v interface{}, k string) (ws []string, errors []error) {
+	value := v.(images.ImageVisibility)
+	if value == images.ImageVisibilityPublic || value == images.ImageVisibilityPrivate || value == images.ImageVisibilityShared || value == images.ImageVisibilityCommunity {
+		return
+	}
+
+	errors = append(errors, fmt.Errorf("%q must be one of %q, %q, %q, %q", k, images.ImageVisibilityPublic, images.ImageVisibilityPrivate, images.ImageVisibilityCommunity, images.ImageVisibilityShared))
+	return
+}
+
+func validatePositiveInt(v interface{}, k string) (ws []string, errors []error) {
+	value := v.(int)
+	if value > 0 {
+		return
+	}
+	errors = append(errors, fmt.Errorf("%q must be a positive integer", k))
+	return
+}
+
+var DiskFormats = [9]string{"ami", "ari", "aki", "vhd", "vmdk", "raw", "qcow2", "vdi", "iso"}
+
+func resourceImagesImagevalidateDiskFormat(v interface{}, k string) (ws []string, errors []error) {
+	value := v.(string)
+	for i := range DiskFormats {
+		if value == DiskFormats[i] {
+			return
+		}
+	}
+	errors = append(errors, fmt.Errorf("%q must be one of %v", k, DiskFormats))
+	return
+}
+
+var ContainerFormats = [9]string{"ami", "ari", "aki", "bare", "ovf"}
+
+func resourceImagesImageValidateContainerFormat(v interface{}, k string) (ws []string, errors []error) {
+	value := v.(string)
+	for i := range ContainerFormats {
+		if value == ContainerFormats[i] {
+			return
+		}
+	}
+	errors = append(errors, fmt.Errorf("%q must be one of %v", k, ContainerFormats))
+	return
+}
+
+func resourceImagesImageV2VisibitlityFromString(v string) images.ImageVisibility {
+	switch v {
+	case "public":
+		return images.ImageVisibilityPublic
+	case "private":
+		return images.ImageVisibilityPrivate
+	case "shared":
+		return images.ImageVisibilityShared
+	case "community":
+		return images.ImageVisibilityCommunity
+	}
+
+	return ""
+}
+
+func resourceImagesImageV2Create(d *schema.ResourceData, meta interface{}) error {
+	config := meta.(*Config)
+	imageClient, err := config.imageV2Client(d.Get("region").(string))
+	if err != nil {
+		return fmt.Errorf("Error creating OpenStack image client: %s", err)
+	}
+
+	protected := d.Get("protected").(bool)
+	createOpts := &images.CreateOpts{
+		Name:            d.Get("name").(string),
+		ContainerFormat: d.Get("container_format").(string),
+		DiskFormat:      d.Get("disk_format").(string),
+		MinDisk:         d.Get("min_disk_gb").(int),
+		MinRAM:          d.Get("min_ram_mb").(int),
+		Protected:       &protected,
+	}
+
+	if v := d.Get("visibility").(string); v != "" {
+		vis := resourceImagesImageV2VisibitlityFromString(v)
+		if vis == "" {
+			return fmt.Errorf("Error unkown image visibility %q", v)
+		}
+		createOpts.Visibility = &vis
+	}
```

Comment:
A few notes about this one:

You can safely omit this entire block because you provided a default value in the schema. Since there is a default value, `v` will never be an empty string.

Instead, just add this to the above `CreateOpts` like so:

```
Visibility: resourceImagesImageV2VisibilityFromString(d.Get("visibility").(string))
```

Or, add this above `CreateOpts`:

```
visibility := resourceImagesImageV2VisibilityFromString(d.Get("visibility").(string))
createOpts := &images.CreateOpts{
    ...
    Visibility: visibility,
}
```

Also, there's no need to return an error if the string is empty. This is for two reasons:

1. You provided a `ValidateFunc`, so an empty string will never be valid and this part of the code will never be reached.
2. In [Gophercloud](https://github.com/gophercloud/gophercloud/blob/master/openstack/imageservice/v2/images/requests.go#L75) the `omitempty` means to ignore an empty string.